File: ios/AidenOnTheGo/LiveActivities/AidenDeepLink.swift
import Foundation

struct AidenNavigationRequest: Equatable, Sendable {
    enum Destination: Equatable, Sendable { case newChat, chat(String) }
    let destination: Destination
    let instanceId: String?
    let workspaceId: String?
    let startsVoice: Bool
}

enum AidenDeepLink {
    static var scheme: String {
        Bundle.main.object(forInfoDictionaryKey: "AidenURLScheme") as? String ?? "aiden-otg"
    }

    static var newChatURL: URL? { newChatURL(instanceId: nil, workspaceId: nil, startsVoice: false) }
    static var newChatVoiceURL: URL? { newChatURL(instanceId: nil, workspaceId: nil, startsVoice: true) }

    static func newChatURL(instanceId: String?, workspaceId: String?, startsVoice: Bool) -> URL? {
        guard instanceId.map(safeID) ?? true, workspaceId.map(safeID) ?? true else { return nil }
        var components = URLComponents()
        components.scheme = scheme
        components.host = startsVoice ? "new-chat-voice" : "new-chat"
        components.queryItems = [
            instanceId.map { URLQueryItem(name: "instance", value: $0) },
            workspaceId.map { URLQueryItem(name: "workspace", value: $0) },
        ].compactMap { $0 }
        return components.url
    }

    static func chatURL(instanceId: String, chatId: String) -> URL? {
        guard safeID(instanceId), safeID(chatId) else { return nil }
        var components = URLComponents()
        components.scheme = scheme
        components.host = "chat"
        components.queryItems = [
            URLQueryItem(name: "instance", value: instanceId),
            URLQueryItem(name: "chat", value: chatId),
        ]
        return components.url
    }

    static func request(from url: URL) -> AidenNavigationRequest? {
        guard url.scheme?.lowercased() == scheme.lowercased(),
              url.user == nil, url.password == nil, url.port == nil,
              url.fragment == nil, url.path.isEmpty else { return nil }
        let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
        guard Set(items.map(\.name)).count == items.count,
              items.allSatisfy({ $0.value != nil }),
              items.allSatisfy({ $0.name == "instance" || $0.name == "workspace" || $0.name == "chat" }) else {
            return nil
        }
        func value(_ name: String) -> String? { items.first(where: { $0.name == name })?.value }
        let instance = value("instance")
        let workspace = value("workspace")
        guard instance.map(safeID) ?? true, workspace.map(safeID) ?? true else { return nil }
        switch url.host?.lowercased() {
        case "new-chat", "new-chat-voice":
            guard value("chat") == nil else { return nil }
            return AidenNavigationRequest(
                destination: .newChat,
                instanceId: instance,
                workspaceId: workspace,
                startsVoice: url.host?.lowercased() == "new-chat-voice"
            )
        case "chat":
            guard let instance, let chat = value("chat"), safeID(chat), workspace == nil else { return nil }
            return AidenNavigationRequest(
                destination: .chat(chat), instanceId: instance, workspaceId: nil, startsVoice: false
            )
        default:
            return nil
        }
    }

    private static func safeID(_ value: String) -> Bool {
        !value.isEmpty && value.count <= 160
            && value.unicodeScalars.allSatisfy { scalar in
                CharacterSet.alphanumerics.contains(scalar) || "._:-".unicodeScalars.contains(scalar)
            }
    }
}

File: android/app/src/main/java/sbtbiswas/AidenOnTheGo/notifications/AidenDeepLink.kt
package sbtbiswas.AidenOnTheGo.notifications

import java.net.URI
import java.net.URLDecoder

sealed class AidenNavigationDestination {
    object NewChat : AidenNavigationDestination()
    data class Chat(val chatId: String) : AidenNavigationDestination()
}

data class AidenNavigationRequest(
    val destination: AidenNavigationDestination,
    val instanceId: String? = null,
    val workspaceId: String? = null,
    val startsVoice: Boolean = false
)

object AidenDeepLink {
    const val SCHEME = "aiden-otg"

    fun newChatUrl(instanceId: String? = null, workspaceId: String? = null, startsVoice: Boolean = false): String {
        val host = if (startsVoice) "new-chat-voice" else "new-chat"
        val params = mutableListOf<String>()
        instanceId?.let { params.add("instance=$it") }
        workspaceId?.let { params.add("workspace=$it") }
        return if (params.isEmpty()) {
            "$SCHEME://$host"
        } else {
            "$SCHEME://$host?${params.joinToString("&")}"
        }
    }

    fun chatUrl(instanceId: String, chatId: String): String {
        return "$SCHEME://chat?instance=$instanceId&chat=$chatId"
    }

    fun parse(uriString: String): AidenNavigationRequest? {
        return try {
            val uri = URI(uriString)
            if (uri.scheme?.lowercase() != SCHEME) return null
            val host = uri.host?.lowercase() ?: return null

            val query = uri.query ?: ""
            val queryMap = query.split("&").filter { it.contains("=") }.associate {
                val parts = it.split("=", limit = 2)
                URLDecoder.decode(parts[0], "UTF-8") to URLDecoder.decode(parts[1], "UTF-8")
            }

            val instanceId = queryMap["instance"]
            val workspaceId = queryMap["workspace"]
            val chatId = queryMap["chat"]

            when (host) {
                "new-chat" -> AidenNavigationRequest(AidenNavigationDestination.NewChat, instanceId, workspaceId, false)
                "new-chat-voice" -> AidenNavigationRequest(AidenNavigationDestination.NewChat, instanceId, workspaceId, true)
                "chat" -> {
                    if (chatId != null && instanceId != null) {
                        AidenNavigationRequest(AidenNavigationDestination.Chat(chatId), instanceId, null, false)
                    } else null
                }
                else -> null
            }
        } catch (_: Exception) {
            null
        }
    }
}
